kore/src/Kore/Log/DebugRewriteTrace.hs

```diff
@@ -0,0 +1,267 @@
+    helpDoc _ = "log every claim before proving it"
+
+instance Entry DebugInitialPattern where
+    entrySeverity _ = Debug
+    oneLineDoc (DebugInitialPattern _) = "Initial pattern"
+    helpDoc _ = "log initial pattern before rewriting"
+
+instance Entry DebugFinalPatterns where
+    entrySeverity _ = Debug
+    oneLineDoc (DebugFinalPatterns _) = "Final patterns"
+    helpDoc _ = "log final patterns after rewriting"
+
+instance Entry DebugRewriteTrace where
+    entrySeverity _ = Debug
+    oneLineDoc DebugRewriteTrace{} = "Rewrite trace"
+    helpDoc _ = "log rewrite substitutions"
+
+unparseOneLine :: Unparse p => p -> Text
+unparseOneLine = renderText . layoutOneLine . unparse
+
+debugInitialClaim ::
+    MonadLog log =>
+    UniqueId ->
+    TermLike VariableName ->
+    log ()
+debugInitialClaim uniqueId claimPattern = logEntry $ DebugInitialClaim uniqueId claimPattern
+
+debugInitialPattern ::
+    MonadLog log =>
+    TermLike VariableName ->
+    log ()
+debugInitialPattern = logEntry . DebugInitialPattern
+
+debugFinalPatterns ::
+    MonadLog log =>
+    MultiOr (Pattern RewritingVariableName) ->
+    log ()
+debugFinalPatterns = logEntry . DebugFinalPatterns . (getRewritingPattern <$>) . from
+
+debugRewriteTrace ::
+    MonadLog log =>
+    UnifyingRuleVariable rule ~ RewritingVariableName =>
+    From rule UniqueId =>
+    Pattern RewritingVariableName ->
+    Results rule ->
+    log ()
+debugRewriteTrace initial Result.Results{results = (toList -> results), remainders} =
+    unless (null results) $
+        logEntry
+            DebugRewriteTrace
+                { initialPattern = getRewritingPattern initial
+                , rewriteResults = getResult <$> results
+                , remainders = multiOrToList remainders
+                }
+  where
+    mapSubstitutionVariables = Substitution.mapVariables (pure toVariableName)
+
+    multiOrToList = (getRewritingPattern <$>) . from
+
+    getResult Result.Result{appliedRule, result} =
+        RewriteResult
+            { ruleId = from @_ @UniqueId $ extract appliedRule
+            , substitution = mapSubstitutionVariables $ Conditional.substitution appliedRule
+            , results = multiOrToList result
+            }
+
+rewriteTraceLogger ::
+    Applicative m =>
+    LogAction m ByteString ->
+    LogAction m SomeEntry
+rewriteTraceLogger textLogger =
+    LogAction action
+  where
+    action entry
+        | Just initial <- fromEntry entry =
+            unLogAction textLogger $ "---\n" <> encode (initial :: DebugInitialClaim) <> "steps:"
+        | Just initial <- fromEntry entry =
+            unLogAction textLogger $ encode (initial :: DebugInitialPattern) <> "steps:"
+        | Just final <- fromEntry entry =
+            unLogAction textLogger $ encode (final :: DebugFinalPatterns)
+        | Just rewrite <- fromEntry entry =
+            unLogAction textLogger $ encode [rewrite :: DebugRewriteTrace]
```

Comment:
(nit) maybe specify the type where `fromEntry` is called instead of later in the body?
Like so:
```suggestion
        | Just initial <- fromEntry @DebugInitialClaim entry =
            unLogAction textLogger $ "---\n" <> encode initial <> "steps:"
        | Just initial <- fromEntry @DebugInitialPattern entry =
            unLogAction textLogger $ encode initial <> "steps:"
...
```